QuickSwitch Compat Q to U

MrSluffy opened the pull request:
## Description
Quickswitch compatibility in Q to U

Special Thanks @liu-wanshun : https://github.com/liu-wanshun/LauncherC

## Type of change
<!-- Replace :x: with :white_check_mark: to "check" the specified bullet -->

:x: General change (non-breaking change that doesn't fit the below categories like copyediting)
:x: Bug fix (non-breaking change which fixes an issue)
:white_check_mark: New feature (non-breaking change which adds functionality)
:x: Breaking change (fix or feature that would cause existing functionality to not work as expected)

MrSluffy committed:
Add Initial support for T

MrSluffy committed:
Merge branch '14-dev' into quickswitch-compat-r-u

MrSluffy committed:
Fixed App to Overview in T

Related changes 3f782edae1d4b0a24aec6c6ad685b3fe8d14b39e

MrSluffy committed:
Fixed crash when taskbar is enabled in T

MrSluffy committed:
Add Initial QS support Q to S

MrSluffy committed:
Fix styling

MrSluffy committed:
Reformat Fixes

MrSluffy committed:
Add compatlibVU

- Seperate compatlib for U for future changes
- Fixed Opening animation in U

MrSluffy committed:
More reformat fixes

- Fixed issue on empty recent on A13 release to qpr2

MrSluffy committed:
Add initial support in A11

- Temp add RemoteTransition, IRemoteTransitionFinishedCallback, TransitionFilter until we create our own implementation

MrSluffy committed:
Fixed gesture unresponsivei in 11 when changing dpi

MrSluffy committed:
Fixed crashes on A10

MrSluffy committed:
Clean up

MrSluffy committed:
Add RemoteTransitionCompat

MrSluffy committed:
temporarily use root screenshot overview in Q

MrSluffy committed:
Fixed NPE

java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.graphics.Bitmap.isRecycled()' on a null object reference
	at android.graphics.BaseCanvas.throwIfCannotDraw(BaseCanvas.java:73)
	at android.graphics.RecordingCanvas.throwIfCannotDraw(RecordingCanvas.java:263)
	at android.graphics.BaseRecordingCanvas.drawBitmap(BaseRecordingCanvas.java:120)
	at com.android.launcher3.graphics.SysUiScrim.draw(SysUiScrim.java:160)
	at com.android.launcher3.LauncherRootView.dispatchDraw(LauncherRootView.java:112)
	at android.view.View.updateDisplayListIfDirty(View.java:22172)
	at android.view.View.draw(View.java:23071)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4619)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4380)
	at android.view.View.updateDisplayListIfDirty(View.java:22172)
	at android.view.View.draw(View.java:23071)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4619)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4380)
	at android.view.View.updateDisplayListIfDirty(View.java:22172)
	at android.view.View.draw(View.java:23071)
	at android.view.ViewGroup.drawChild(ViewGroup.java:4619)
	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4380)
	at android.view.View.draw(View.java:23361)
	at com.android.internal.policy.DecorView.draw(DecorView.java:847)
	at android.view.View.updateDisplayListIfDirty(View.java:22181)
	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:682)
	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:688)
	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:786)
	at android.view.ViewRootImpl.draw(ViewRootImpl.java:4885)
	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:4584)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3787)
	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2514)
	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:9510)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1326)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1334)
	at android.view.Choreographer.doCallbacks(Choreographer.java:967)
	at android.view.Choreographer.doFrame(Choreographer.java:881)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1309)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:204)
	at android.os.Looper.loop(Looper.java:291)
	at android.app.ActivityThread.main(ActivityThread.java:8133)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:588)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1019)

MrSluffy committed:
Fixed not trigger screenshot in 11

MrSluffy committed:
Fixed NME causing empty recent

java.lang.NoSuchMethodError: No interface method toList()Ljava/util/List; in class Ljava/util/stream/Stream; or its super classes (declaration of 'java.util.stream.Stream' appears in /apex/com.android.art/javalib/core-oj.jar)
	at com.android.quickstep.TopTaskTracker$CachedTaskInfo.otherVisibleTaskThisIsExcludedOver(TopTaskTracker.java:256)

MrSluffy committed:
Convert java code to kt

MrSluffy committed:
Revert "Convert java code to kt"

This reverts commit 85f110bc7e0b4b317dc4552d3e3b65908988d58d.

MrSluffy committed:
Fixed crash oos12

MrSluffy committed:
Fixed all apps menu in taskbar

MrSluffy committed:
Fix  java.lang.NoClassDefFoundError in A13 below

java.lang.NoClassDefFoundError: Failed resolution of: Landroid/window/IDumpCallback$Stub;
	at com.android.app.viewcapture.SettingsAwareViewCapture.getInstance(Unknown Source:0)
	at com.android.launcher3.taskbar.overlay.TaskbarOverlayDragLayer.onAttachedToWindow(TaskbarOverlayDragLayer.java:83)
	at android.view.View.dispatchAttachedToWindow(View.java:21357)
	at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3491)
	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2900)

MrSluffy committed:
Fixed crash on oos 12 when toping outside card

MrSluffy committed:
Fixed crash task menu bar when work profile is enabled

MrSluffy committed:
Convert compatlib java to kt

MrSluffy committed:
Merge branch '14-dev' into quickswitch-compat-r-u

MrSluffy committed:
Fixed build

MrSluffy committed:
Fixed empty recent in some oneui core 5.1

MrSluffy committed:
Revert "Convert compatlib java to kt" as causes gesture issue in oneui

This reverts commit 73c3ab1bce00c3c72a6b2dc459519f539235efac.

MrSluffy committed:
Fixed flickering when tapping nav bar

 repro : Open app -> tap nav : flick starting

MrSluffy committed:
Add readme

Goooler committed:
Clean up deps

Goooler committed:
Clean up deps

MrSluffy committed:
Add support legacy split screen

Co-Authored-By: 无言 <57122860+liu-wanshun@users.noreply.github.com>

MrSluffy committed:
Fixed NoSuchMethodError in OOS 13

Lawnchair (Debug) bug report 22 янв. 2024 г. 17:33:14
version: 14 Dev (#363) (14)
commit: 1f3491f
build.brand: OnePlus
build.device: OnePlus9
build.display: TQ2A.230505.002
build.fingerprint: OnePlus/lemonade/lemonade:13/TQ2A.230505.002/1241:user/release-keys
build.hardware: qcom
build.id: TQ2A.230505.002
build.manufacturer: OnePlus
build.model: LE2110
build.security.level: 2023-05-05
build.product: OnePlus9
build.type: user
version.codename: REL
version.incremental: 1686400879
version.release: 13
version.sdk_int: 33
display.density_dpi: 443
isRecentsEnabled: true

error: Uncaught exception

java.lang.NoSuchMethodError: No virtual method registerForCallingUser()V in class Lcom/android/internal/policy/GestureNavigationSettingsObserver; or its super classes (declaration of 'com.android.internal.policy.GestureNavigationSettingsObserver' appears in /system/framework/framework.jar!classes4.dex)
	at com.android.launcher3.taskbar.TaskbarInsetsController.init(TaskbarInsetsController.kt:80)
	at com.android.launcher3.taskbar.TaskbarControllers.init(TaskbarControllers.java:168)
	at com.android.launcher3.taskbar.TaskbarActivityContext.init(TaskbarActivityContext.java:313)
	at com.android.launcher3.taskbar.TaskbarManager.recreateTaskbar(TaskbarManager.java:460)
	at com.android.launcher3.taskbar.TaskbarManager.onUserUnlocked(TaskbarManager.java:364)
	at com.android.quickstep.TouchInteractionService$$ExternalSyntheticLambda7.run(Unknown Source:2)
	at com.android.launcher3.util.RunnableList.add(RunnableList.java:36)
	at com.android.launcher3.util.LockedUserState.runOnUserUnlocked(LockedUserState.kt:72)
	at com.android.quickstep.TouchInteractionService.onCreate(TouchInteractionService.java:506)
	at android.app.ActivityThread.handleCreateService(ActivityThread.java:4494)
	at android.app.ActivityThread.-$$Nest$mhandleCreateService(Unknown Source:0)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2158)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7891)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:550)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)

MrSluffy committed:
Re-Fixed empty recent

MrSluffy committed:
Fixed IllegalArgumentException: width & height must be > 0

Lawnchair (Debug) bug report Jan 23, 2024 01:35:38
version: 14 Dev (#365) (14)
commit: ab25769
build.brand: samsung
build.device: b2q
build.display: TP1A.220624.014.F711BXXS6FWK2
build.fingerprint: samsung/b2qxeea/b2q:13/TP1A.220624.014/F711BXXS6FWK2:user/release-keys
build.hardware: qcom
build.id: TP1A.220624.014
build.manufacturer: samsung
build.model: SM-F711B
build.security.level: 2023-11-01
build.product: b2qxeea
build.type: user
version.codename: REL
version.incremental: F711BXXS6FWK2
version.release: 13
version.sdk_int: 33
display.density_dpi: 391
isRecentsEnabled: true

error: Uncaught exception

java.lang.RuntimeException: Unable to start activity ComponentInfo{app.lawnchair.debug/app.lawnchair.LawnchairLauncher}: java.lang.IllegalArgumentException: width & height must be > 0
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4184)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4340)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2584)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.app.ActivityThread.main(ActivityThread.java:8810)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:604)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
Caused by: java.lang.IllegalArgumentException: width & height must be > 0
	at android.graphics.Bitmap.createBitmap(Bitmap.java:1300)
	at android.graphics.Bitmap.createBitmap(Bitmap.java:1277)
	at com.android.launcher3.icons.BitmapRenderer.createHardwareBitmap(BitmapRenderer.java:52)
	at com.android.launcher3.icons.ShadowGenerator$Builder.createPill(ShadowGenerator.java:151)
	at com.android.launcher3.icons.ShadowGenerator$Builder.createPill(ShadowGenerator.java:138)
	at com.android.launcher3.icons.DotRenderer.<init>(DotRenderer.java:78)
	at com.android.launcher3.DeviceProfile.createDotRenderer(DeviceProfile.java:726)
	at com.android.launcher3.DeviceProfile.<init>(DeviceProfile.java:708)
	at com.android.launcher3.DeviceProfile$Builder.build(DeviceProfile.java:2255)
	at com.android.launcher3.InvariantDeviceProfile.initGrid(InvariantDeviceProfile.java:470)
	at com.android.launcher3.InvariantDeviceProfile.initGrid(InvariantDeviceProfile.java:359)
	at com.android.launcher3.InvariantDeviceProfile.initGrid(InvariantDeviceProfile.java:373)
	at com.android.launcher3.InvariantDeviceProfile.<init>(InvariantDeviceProfile.java:233)
	at com.android.launcher3.InvariantDeviceProfile.$r8$lambda$DNcXzmawjoq65q3wgQi9M48DryY(Unknown Source:2)
	at com.android.launcher3.InvariantDeviceProfile$$ExternalSyntheticLambda4.get(Unknown Source:0)
	at com.android.launcher3.util.MainThreadInitializedObject.lambda$get$0(MainThreadInitializedObject.java:60)
	at com.android.launcher3.util.MainThreadInitializedObject.$r8$lambda$xtvr3aByNYHC2nsBQ97HC1SKg4A(Unknown Source:0)
	at com.android.launcher3.util.MainThreadInitializedObject$$ExternalSyntheticLambda0.get(Unknown Source:4)
	at com.android.launcher3.util.TraceHelper.allowIpcs(TraceHelper.java:100)
	at com.android.launcher3.util.MainThreadInitializedObject.get(MainThreadInitializedObject.java:59)
	at com.android.launcher3.LauncherAppState.<init>(LauncherAppState.java:165)
	at com.android.launcher3.LauncherAppState.<init>(LauncherAppState.java:111)
	at com.android.launcher3.LauncherAppState$$ExternalSyntheticLambda16.get(Unknown Source:2)
	at com.android.launcher3.util.MainThreadInitializedObject.lambda$get$0(MainThreadInitializedObject.java:60)
	at com.android.launcher3.util.MainThreadInitializedObject.$r8$lambda$xtvr3aByNYHC2nsBQ97HC1SKg4A(Unknown Source:0)
	at com.android.launcher3.util.MainThreadInitializedObject$$ExternalSyntheticLambda0.get(Unknown Source:4)
	at com.android.launcher3.util.TraceHelper.allowIpcs(TraceHelper.java:100)
	at com.android.launcher3.util.MainThreadInitializedObject.get(MainThreadInitializedObject.java:59)
	at com.android.launcher3.LauncherAppState.getInstance(LauncherAppState.java:99)
	at com.android.launcher3.Launcher.onCreate(Launcher.java:519)
	at com.android.launcher3.uioverrides.QuickstepLauncher.onCreate(QuickstepLauncher.java:578)
	at app.lawnchair.LawnchairLauncher.onCreate(LawnchairLauncher.kt:215)
	at android.app.Activity.performCreate(Activity.java:8657)
	at android.app.Activity.performCreate(Activity.java:8636)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1417)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4165)
	... 12 more

Goooler committed:
Apply AOSP style for Java files

Goooler committed:
Move src dirs

Goooler commented on compatLib/build.gradle:
```diff
@@ -4,6 +4,17 @@ plugins {
+        aidl true
+    }
+
+    sourceSets {
+        main {
+            java.srcDirs = ["src"]
+            aidl.srcDirs = ['src']
+        }
+    }
```
Is there any specific reason to move these default source sets?

MrSluffy replied:
No, does this causing any issue?

Goooler replied:
Nope, if not, I suggest we can put them back.

MrSluffy committed:
Fix security exception

Lawnchair (Debug) bug report Jan 24, 2024 11:10:46 AM
version: 14 Dev (#368) (14)
commit: cde8994
build.brand: POCO
build.device: alioth
build.display: UQ1A.240105.004 release-keys
build.fingerprint: POCO/alioth_global/alioth:13/TKQ1.220829.002/V14.0.4.0.TKHMIXM:user/release-keys
build.hardware: qcom
build.id: UQ1A.240105.004
build.manufacturer: Xiaomi
build.model: M2012K11AG
build.security.level: 2024-01-05
build.product: lineage_alioth
build.type: user
version.codename: REL
version.incremental: 1705128613
version.release: 14
version.sdk_int: 34
display.density_dpi: 420
isRecentsEnabled: true

error: Uncaught exception

java.lang.SecurityException: Permission Denial: createSearchSession from pid=7829, uid=10727
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3061)
	at android.os.Parcel.createException(Parcel.java:3045)
	at android.os.Parcel.readException(Parcel.java:3028)
	at android.os.Parcel.readException(Parcel.java:2970)
	at android.app.search.ISearchUiManager$Stub$Proxy.createSearchSession(ISearchUiManager.java:219)
	at android.app.search.SearchSession.<init>(SearchSession.java:110)
	at android.app.search.SearchUiManager.createSearchSession(SearchUiManager.java:54)
	at app.lawnchair.search.LawnchairDeviceSearchAlgorithm$Companion.checkSearchCompatibility$lambda$2(LawnchairDeviceSearchAlgorithm.kt:128)
	at app.lawnchair.search.LawnchairDeviceSearchAlgorithm$Companion.$r8$lambda$Xo_kUeSrgktAFvrDmVpYiu_nWd4(Unknown Source:0)
	at app.lawnchair.search.LawnchairDeviceSearchAlgorithm$Companion$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.os.HandlerThread.run(HandlerThread.java:67)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.searchui.SearchUiManagerService$SearchUiManagerStub.runForUserLocked(SearchUiManagerService.java:173)
	at com.android.server.searchui.SearchUiManagerService$SearchUiManagerStub.createSearchSession(SearchUiManagerService.java:109)
	at android.app.search.ISearchUiManager$Stub.onTransact(ISearchUiManager.java:126)
	at android.os.Binder.execTransactInternal(Binder.java:1358)
	at android.os.Binder.execTransact(Binder.java:1304)

MrSluffy committed:
Fixed glitches multi tapping nav pill in U

MrSluffy committed:
Merge branch '14-dev' into quickswitch-compat-r-u

Goooler committed:
Move source sets back

Goooler committed:
Replace compat lib paths
